### ISSUE TITLE
Unable to get Topic Information when connecting to a Cluster using SSL

### ISSUE THREAD

**Issue by jheinitz** [StephenSorriaux/ansible-kafka-admin#128]
Hello,

first of all, I would like to thank you for this brilliant project. We are starting to implement Kafka and want to manage topics through Ansible. That said, this project fits perfectly. However, I just faced an issue, which might already be referenced in #123, but my error is slightly different. Here are the details:

## Expected Behavior
run task kafka_info with resource: "topic" and get a list of all available topics

## Actual Behavior
Ansible playbook stops with an error:

## Additional information
1. The same playbooks can successfully retrieve the `broker` and the `topic-config` through the kafka_info module. It only fails when I want to retrieve the `topic` information. 
2. The CN of the certificate is a super user of the Kafka Cluster.

## Play to Reproduce the Problem
```
---
- hosts: kafka11t
  remote_user: root
  vars:
    # see roles/cert/default/main.yml for all available
    # parameters
    my_bootstrap_servers: "[kafka11t.example.com:9093](http://kafka11t.example.com:9093/),[kafka12t.example.com:9093](http://kafka12t.example.com:9093/),[kafka13t.example.com:9093](http://kafka13t.example.com:9093/),[kafka14t.example.com:9093](http://kafka14t.example.com:9093/)"
    my_zookeeper_servers: "[kafka11t.example.com:2182](http://kafka11t.example.com:2182/)"
    kafka_api_version: "2.6.0"

  tasks:
    # get information about topics
    - name: get topic information
      StephenSorriaux.ansible_kafka_admin.kafka_info:
        resource: "topic"
        bootstrap_servers: "{{ my_bootstrap_servers }}"
        api_version: "{{ kafka_api_version }}"
        security_protocol: "SSL"
        ssl_cafile: "/etc/pki/ca-trust/source/anchors/MMW-CA-SHA256.cer.crt"
        ssl_certfile: "/etc/pki/tls/certs/{{ ansible_fqdn }}.crt"
        ssl_keyfile: "/etc/pki/tls/private/{{ ansible_fqdn }}.pem"
      changed_when: false
      register: topics_info

    - name: "Display information for topic jh03"
      debug:
        var: topics_info['ansible_module_results']['jh03']
...

## Logs from the play with Ansible in debug mode

`ANSIBLE_DEBUG=true ansible-playbook my-awesome-playbook.yml`

```
[WARNING]: Module invocation had junk after the JSON data: <BrokerConnection node_id=2 host=[kafka12t.example.com:9093](http://kafka12t.example.com:9093/) <connected> [IPv4 ('10.100.210.63', 9093)]>: Closing connection.  <BrokerConnection node_id=4
host=[kafka14t.example.com:9093](http://kafka14t.example.com:9093/) <connecting> [IPv4 ('10.100.210.72', 9093)]>: Closing connection.
fatal: [kafka11t]: FAILED! => {"changed": false, "msg": "Error while getting topic from Kafka: KafkaManagerError: Connection is not ready, please check your client and server configurations. "}
```

## Specifications

  - Library version: 0.15.3
  - Result of `pip list` command: 
  - Kafka version:  Confluent Platform 7.0.1
  - Python version: 3.6.8
  - OS: Linux - RHEL 8.4 on Ansible Node

```

**Comment by jheinitz** [comment]
Hello,

I just recognized that I get exactly the same error message when I try to get information about consumer groups using the resource `consumer_group` in the module kafka_info. 

Thank you in advance and kind regards

Jens

**Comment by StephenSorriaux** [comment]
Hello,

Thank you for the issue and the kind words. Glad to read this lib can be useful to others!

I think you might be right mentionning #123, which was linked to some missing ACLs. Moreover the fact that you can retrieve information using the ‘broker’ parameter seems to confirm that, and also validates your SSL configuration is correct.

I am wondering if the issue is due to the fact that you are contacting the SSL listener (on port 9093/tcp) which, in my understanding, does not provide any authentication « layer ». Do you have any way to check your Kafka configuration and see if any SASL_SSL listener is available (from Confluent documentation, it might be on available on port 9094/tcp)? If so, can you retry your playbook using the SASL_SSL listener and a Kafka user with enough rights?

**Comment by jheinitz** [comment]
Hi Stephen!

Thanks for your reply. In the meantime, I can give you a quick update what I did so far:
1. I downloaded the zip of the master branch and create an ansible collection using the script you provided. I installed that one and re-ran my playbook. The error is the same. So this seem to proove that #123 is not the root cause for my issue.
2. The authentication layer on port 9093 (SSL) is performed by checking the CN of the certificate. We have a rule to extract the CN= from the DN and strip some unneeded information. Basically it works like this: a) find CN=<something>.example.com<the unneeded rest> b) extract <something> from the CN= string, c) use "something" as the User to connect to the Kafka Cluster. The cluster is configured with a user "something" to be one of the Cluster's super user. This works as expected, because when I try to connect with a certificate that has a CN=notvalid.wxample.com, I cannot connect to the cluster at all. 
3. With the same setup and connection to port 9093 I am able to connect to the cluster using kafka-topics, kafka-config and/or kafka-acls. I can also connect to a cluster using AKHQ.
4. I will give it a try to connect to port 9094, but we use OAuthbearer on that port. Not sure if this works with your lib.

Could it be an option to dump "the junk after the JSON data" to a temporary file to see what's wrong here?

Thanks and kind regards

Jens

**Comment by jheinitz** [comment]
Hi! 

It's me again. I digged deeper into the kafka_manager.py file and figured out that the problem occurs during the call to `send_request_and_get_response()` in the "for node in brokers" loop at around line 1249. 

Another strange thing is that the problem does not occur, when I stop one of my brokers. I have a setup of four brokers and most of my topics have four replicas. 

Maybe this gives you a hint and helps you to adress the issue. Atleast we can say that it is not a permission issue, because I get the information on the topics when I comment the lines 1224-1279.

Looking forward to hear from you.

Thanks and best regards

Jens

**Comment by StephenSorriaux** [comment]
Hello,

Oh, I see, you are using mTLS, thanks for the explanation. I don't think this lib supports Oauth, but it can be added (https://github.com/dpkp/kafka-python/pull/1750).

It should be possible to dump the logs produced by the kafka client we are using to a file by changing this line (or adding a similar line just after): https://github.com/StephenSorriaux/ansible-kafka-admin/blob/master/module_utils/ssl_utils.py#L9 using a `FileHandler` instead (https://docs.python.org/3/library/logging.handlers.html#filehandler)

Thanks for digging more, it seems like the issue is with the `OffsetRequest` that may not be correct or getting refused by the server. Do you have any logs from your Kafka servers when the issue occurs?

**Comment by StephenSorriaux** [comment]
I see Confluent Platform 7 is based on Kafka 3.0.X, I am not sure we did test the lib against this version yet, but since you got it working with 3 brokers, I believe that should not be a problem.

The 4 brokers vs 3 brokers test you mentioned is kind of weird. Are you using kRaft or Zookeeper? As you are using docker containers, would it possible for you to share your current setup (here or in a gist) so that I can perform some tests locally?

**Comment by jheinitz** [comment]
Hi!

We are using Confluent Platform 7.0.1, but we are still using Zookeepers. I 
will try to create a setup without our AD integration and the OAuthbearer 
configuration on port 9093. Then we can talk about sharing the config.

I will come back to you tomorrow.

Thanks and best regards

Jens

Am 24. Februar 2022 20:26:49 schrieb Stephen Sorriaux 
***@***.***>:

**Comment by ryarnyah** [comment]
@jheinitz Some questions to better understand how we can find a solution to your issue:
* Could you share with us your configuration (without secrets & private data of course)? (if you don't want it to be in github we can exchange it on another canal)
* Do you have the same version on all your brokers?
* Is there some firewall between you and the brokers?
* Do you have some warn or error in your brokers logs when problem occurred?

**Comment by jheinitz** [comment]
Hello!

* we are running Kafka in docker Containers. I could get the generated kafka.properties files form the containers and share these here. 
* yes, we are using Confluent Platform 7.1.1
* There is a firewall between the Ansible Controller and the Kafka Brokers.
* no, not at all.

Jens

**Comment by jheinitz** [comment]
Hi, please find my kafka.properties files for the four brokers in my cluster attached.

[kafka.properties_kafkabr4test.txt](https://github.com/StephenSorriaux/ansible-kafka-admin/files/8909380/kafka.properties_kafkabr4test.txt)
[kafka.properties_kafkabr1test.txt](https://github.com/StephenSorriaux/ansible-kafka-admin/files/8909381/kafka.properties_kafkabr1test.txt)
[kafka.properties_kafkabr2test.txt](https://github.com/StephenSorriaux/ansible-kafka-admin/files/8909382/kafka.properties_kafkabr2test.txt)
[kafka.properties_kafkabr3test.txt](https://github.com/StephenSorriaux/ansible-kafka-admin/files/8909383/kafka.properties_kafkabr3test.txt)

**Comment by teebee** [comment]
Recently, I used the task `kafka_info` with `resource: topic` for the first time and had the same issue as @jheinitz . In a cluster with few topics/partitions I could not observe the problem, in a cluster with many topics/partitions it occurred regularly.

I then took a closer look at the source code. The error occurs in [send_request_and_get_response()](https://github.com/StephenSorriaux/ansible-kafka-admin/blob/master/module_utils/kafka_manager.py#L481)  because [self.connection_check()](https://github.com/StephenSorriaux/ansible-kafka-admin/blob/master/module_utils/kafka_manager.py#L506) fails. After I had set `MAX_RETRY = 60` (see [kafka_manager.py#L78](https://github.com/StephenSorriaux/ansible-kafka-admin/blob/master/module_utils/kafka_manager.py#L78)), I could no longer observe the error. So it just seems to be a timing issue.

@StephenSorriaux @ryarnyah Can you please increase `MAX_RETRY` accordingly or - better - make it configurable?

**Comment by teebee** [comment]
For my purposes, it would be sufficient to change [kafka_manager.py#L78](https://github.com/StephenSorriaux/ansible-kafka-admin/blob/master/module_utils/kafka_manager.py#L78) to
```
    MAX_RETRY = int(os.getenv('MAX_RETRY', '10'))
```
(and of course add `import os`).

**Comment by teebee** [comment]
LGTM. With the new default value of 50, the problem no longer occurs for me. However, I have also tested whether `connect_max_retry` can be changed individually per task. This was also successful. Many thanks for the quick and excellent solution. Now I would be happy if there would be a new version soon...

**Comment by StephenSorriaux** [comment]
Thank you for the feedback @teebee, I just released the `0.20.0` version.

@jheinitz if by any chance you can give it a try, it would be great!

**Comment by jheinitz** [comment]
Hello @StephenSorriaux ,

sorry for the delay, but I was sick. I just gave it a try and I can confirm that it is now working with the resource `topic`. I had to set `connect_max_retry: 25` in order to get results every time.

However, I also tried to get information on consumer group using:
```
    # get information about consumer groups
    - name: get consumer groups information
      StephenSorriaux.ansible_kafka_admin.kafka_info:
        resource: "consumer_group"
        bootstrap_servers: "{{ my_bootstrap_servers }}"
        api_version: "2.6.0"
        security_protocol: "SSL"
        ssl_cafile: "/etc/pki/ca-trust/source/anchors/MMW-CA-SHA256.cer.crt"
        ssl_certfile: "/etc/pki/tls/certs/{{ ansible_fqdn }}.crt"
        ssl_keyfile: "/etc/pki/tls/private/{{ ansible_fqdn }}.pem"
      changed_when: false
      register: groups_info
```
but all I get is this error:
```
 ________________________________________
< TASK [get consumer groups information] >
 ----------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

[WARNING]: Module invocation had junk after the JSON data: <BrokerConnection node_id=3 host=kafkabr3test.example.com:9093 <connected> [IPv4
('192.168.1.85', 9093)]>: Closing connection.
fatal: [kafkabr4test]: FAILED! => {"changed": false, "msg": "Something went wrong: 'utf-8' codec can't decode byte 0xe4 in position 8: invalid continuation byte "}
```
Is this a completely different problem or does it need the same fix?

Kind regards

Jens
